Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,262 @@
+# HVPA-EP-0001: HVPA Redesign
+
+## Summary
+This proposal outlines a plan to enable HVPA to do more efficient scaling horizontally and vertically.
+
+## Motivation
+Currently, HVPA scales by applying user provided weights to HPA and VPA recommendations. We need to provide a better control over scaling of the applications to the users.
+
+### Goals
+
+1. HVPA should be able to scale an application with finer-grain control and predictability over horizontal and vertical scaling which can be specified by the users who are deploying/managing the application.
+
+## Approach
+
+Users will provide buckets of resources each of which would include minimum and maximum CPU, memory and replicas respectively. A configuration is said to belong to a bucket if total resources allocated to all its replcas lie in that bucket. For eg. if there are 3 replicas with 4 cores each, i.e., total 12 cores, then this configuration could belong to a bucket with:
+```
+minimum CPU: 3
+maximum CPU: 6
+minimum memory: 2.5G
+maximum memory: 7G
+minimum replicas: 2
+maximum replicas: 5
+```
+For this bucket mimimum and maximum total CPU are 6 (= 3 * 2) and 30 (= 6 * 5) respectively, and same for the memory would be 5G and 35G respectively.
+So a configuration with 3 replicas with 6GB memory each would also lie in the same bucket.
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+If the aggregated recommendations fall in this bucket, then the size of a replica of the scaled application would always be more than the minimum values of resources provided for the bucket, i.e., 3 cores and 2.5G in this case.
+However, the maximum value may not necessarily be bound by its maximum value. For details refer [this section](####base-usage-of-each-replica). For this particular example, if HPA recommends 4 replicas, and VPA recommends 8G memory, then after the scaling, the memory of each replica would be 8GB even though aggregated memory recommendation fall in this bucket.
```

Comment:
We are not merging the recommendations from HPA and VPA here. I think the confusion here is due to the mention of HPA recommendations. i would/should remove the HPA mention in next commit.
For the example considered here, i.e., for replica count 3, the following updated statement would still mean the same though:
>For this particular example, if VPA recommends 8G memory, then after the scaling, the memory of each replica would be 8GB even though aggregated memory recommendation fall in this bucket.

**Explanation:**
Current: Replicas: 3, memory: anyvalue
VPA recommends 8GB. So total recommendation is 24G, which falls in this bucket.
Scaling according to the bucket strategy:
New replicas: 5
New memory: `24/5 = 4.8G`
But HVPA would set the new memory to 8GB, which would take the total memory to `5 * 8 = 40G`.
Refer chart for calculations:
<img width="512" alt="Screenshot 2020-05-26 at 8 10 32 AM" src="https://user-images.githubusercontent.com/31065672/82855247-62019300-9f28-11ea-873f-40d1ec8596cf.png">

This is as per the example that we discussed over email:
```
Current Requests: 150m / 1.8Gi with 1 replica
VPA Recommendation: 150m / 2.2Gi
New Requests: 150m / 2.2Gi with 1->2 replicas, because 2.2Gi lands us in the “second” bucket when out-scaling that has 2 replicas; we carry over the vertical scaling as is, which is most conservative.
Reasoning: We don’t know what the cache percentage is and if it is very high, the new replica will need almost the same memory to hold the cache. If not, VPA will eventually scale down.
```

This is also explained under section `base usage of each replica`.

This sounded acceptable to do when resource values were small (1.8G, 2.2G), but the scale of overscaling is more apparent when slightly larger (but still practical) values are considered.

Reply:
i will also update the doc with the above explanation to make it more clear.